### PR TITLE
Cbuild Orchestration

### PR DESCRIPTION
**Pull request by soumeh01**
The change is concerning [issue-28](https://github.com/Open-CMSIS-Pack/cbuild/issues/28)

- Taking input csolution.yml file & options `--context` , `--load`, `--output`
- Added commands `list contexts` & `list toolchains`
- extended `--log` option  to other commands & subcommands
- Updated usage and documentation
- Added tests

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -4,32 +4,41 @@
 
 # cbuild: Open-CMSIS-Pack Build Invocation Utility
 
-This utility allows embedded developers to orchestrate the build of CPRJ projects using `cbuildgen`, `cpackget`, `cmake` and `ninja.`
+This utility allows embedded developers to orchestrate the build of CPRJ and CSolution projects using `cbuildgen`, `csolution`, `cpackget`, `cmake` and `ninja.`
```
Comment (jkrech) [inline README.md:7]: This utility allows embedded developers to build CPRJ and csolution projects by orchestrating the following tools: ...

File: README.md [hunk 2]
```diff
@@ -4,32 +4,41 @@
...
   -i, --intdir string      Set directory for intermediate files
   -j, --jobs int           Number of job slots for parallel execution
+      --load string        Set policy for packs loading [latest|all|required]
   -l, --log string         Save output messages in a log file
   -o, --outdir string      Set directory for output files
   -p, --packs              Download missing software packs with cpackget
   -q, --quiet              Suppress output messages except build invocations
   -r, --rebuild            Remove intermediate and output directories and rebuild
-  -s, --schema             Check *.cprj file against CPRJ.xsd schema
+  -s, --schema             Check input file schema
```
Comment (jkrech) [inline README.md:36]: Validate project input file(s) against schema

File: pkg/utils/configs.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+	"os"
+	"path/filepath"
+	"runtime"
+
+	log "github.com/sirupsen/logrus"
+)
+
+type Configurations struct {
+	BinPath string
+	ETCPath string
```
Comment (jkrech) [inline pkg/utils/configs.go:19]: I would expect `EtcPath` here.

**Review by jkrech — COMMENTED** [review]
The updates to the ReadMe in response to my feedback are good.

File: pkg/builder/cproject/builder.go [hunk 4]
```diff
@@ -16,37 +16,14 @@ import (
...
-	Jobs      int
-	Quiet     bool
-	Debug     bool
-	Verbose   bool
-	Clean     bool
-	Schema    bool
-	Packs     bool
-	Rebuild   bool
-	UpdateRte bool
+type CPRJBuilder struct {
```
Comment (brondani) [inline pkg/builder/cproject/builder.go:25]: For consistency with devtools [guidelines](https://github.com/Open-CMSIS-Pack/devtools/blob/main/docs/GUIDELINES.md#source-files-cpp-java) should we apply camel case here `CprjBuilder`?

File: README.md [hunk 5]
```diff
@@ -4,32 +4,47 @@
...
+cbuild: Build Invocation 1.3.0 (C) 2022 Arm Ltd. and Contributors
 
 Usage:
-  cbuild <project.cprj> [flags]
+  cbuild [command] <project.cprj|csolution.yml> [flags]
+
+Available Commands:
+  help            Help about any command
+  list-contexts   Print list of contexts in a csolution.yml
+  list-toolchains Print list of installed toolchains
```
Comment (brondani) [inline README.md:26]:
```suggestion
  list contexts   Print list of contexts in a csolution.yml
  list toolchains Print list of installed toolchains
```

File: pkg/builder/csolution/builder.go [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+
+	csolutionBin := filepath.Join(b.InstallConfigs.BinPath, "csolution"+b.InstallConfigs.BinExtn)
+	if _, err := os.Stat(csolutionBin); os.IsNotExist(err) {
+		log.Error("csolution was not found")
+		return nil, err
+	}
+
+	output, err := b.Runner.ExecuteCommand(csolutionBin, quite, args...)
+	if err != nil {
+		log.Error("error executing 'cbuild list-contexts'")
```
Comment (brondani) [inline pkg/builder/csolution/builder.go:44]:
```suggestion
		log.Error("error executing 'cbuild list contexts'")
```

File: pkg/builder/csolution/builder.go [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+	if _, err := os.Stat(csolutionBin); os.IsNotExist(err) {
+		log.Error("csolution was not found")
+		return toolchains, err
+	}
+
+	args := []string{"list", "toolchains", "--solution=" + b.InputFile}
+
+	output, err := b.Runner.ExecuteCommand(csolutionBin, quite, args...)
+	if err != nil {
+		log.Error("error executing 'cbuild list-toolchains'")
```
Comment (brondani) [inline pkg/builder/csolution/builder.go:65]:
```suggestion
		log.Error("error executing 'cbuild list toolchains'")
```

File: pkg/builder/csolution/builder.go [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+	if output != "" {
+		contexts = strings.Split(strings.ReplaceAll(strings.TrimSpace(output), "\r\n", "\n"), "\n")
+	}
+	return contexts, nil
+}
+
+func (b CSolutionBuilder) listToolchains(quite bool) (toolchains []string, err error) {
+	csolutionBin := filepath.Join(b.InstallConfigs.BinPath, "csolution"+b.InstallConfigs.BinExtn)
+	if _, err := os.Stat(csolutionBin); os.IsNotExist(err) {
+		log.Error("csolution was not found")
```
Comment (brondani) [inline pkg/builder/csolution/builder.go:57]: The `csolution` option in the `list toolchains` command must be optional.

**Review by brondani — COMMENTED** [review]
The `--log` option to capture output messages should be available for all commands, including the new ones.

I would also recommend to increase the coverage of these modules:
```
ok  	cbuild/cmd/cbuild/commands/list	2.019s	coverage: 66.7% of statements
ok  	cbuild/pkg/builder/csolution	2.005s	coverage: 66.0% of statements
```

File: pkg/builder/cproject/builder.go (hunk 4 above)
Comment (soumeh01) [reply pkg/builder/cproject/builder.go]: Addressed in one of the changes done in [Improvements](https://github.com/Open-CMSIS-Pack/cbuild/pull/29/commits/472704065a8867715b9fefc2c651f85f036e9c02)

File: README.md (hunk 5 above)
Comment (soumeh01) [reply README.md]: Addressed in one of the changes done in [Improvements](https://github.com/Open-CMSIS-Pack/cbuild/pull/29/commits/472704065a8867715b9fefc2c651f85f036e9c02)

File: pkg/builder/csolution/builder.go (hunk 6 above)
Comment (soumeh01) [reply pkg/builder/csolution/builder.go]: Addressed in one of the changes done in [Improvements](https://github.com/Open-CMSIS-Pack/cbuild/pull/29/commits/472704065a8867715b9fefc2c651f85f036e9c02)

File: pkg/builder/csolution/builder.go (hunk 7 above)
Comment (soumeh01) [reply pkg/builder/csolution/builder.go]: Addressed in one of the changes done in [Improvements](https://github.com/Open-CMSIS-Pack/cbuild/pull/29/commits/472704065a8867715b9fefc2c651f85f036e9c02)

File: pkg/builder/csolution/builder.go (hunk 8 above)
Comment (soumeh01) [reply pkg/builder/csolution/builder.go:57]: Addressed in one of the changes done in [Improvements](https://github.com/Open-CMSIS-Pack/cbuild/pull/29/commits/472704065a8867715b9fefc2c651f85f036e9c02)

**Review by brondani — APPROVED** [review]

**Pull request merged**